### DISCUSSION TITLE
Introspect naming behavior in a simple message <-> user relation example

### DISCUSSION THREAD

**Discussion by jrmyio** [prisma/prisma#3423] [Q&A] [answered]
The following SQL code:

```sql
SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
SET time_zone = "+00:00";

CREATE TABLE `avatar` (
  `id` int(11) NOT NULL,
  `avatar_url` text COLLATE utf8_bin NOT NULL,
  `user_id` int(11) NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;

CREATE TABLE `message` (
  `id` int(11) NOT NULL,
  `sender_id` int(11) NOT NULL,
  `receiver_id` int(11) NOT NULL,
  `message` text COLLATE utf8_bin NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;

CREATE TABLE `user` (
  `id` int(11) NOT NULL,
  `name` varchar(11) COLLATE utf8_bin NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;

INSERT INTO `user` (`id`, `name`) VALUES
(1, 'test'),
(2, 'test2');


ALTER TABLE `avatar`
  ADD PRIMARY KEY (`id`),
  ADD KEY `user_id` (`user_id`);

ALTER TABLE `message`
  ADD PRIMARY KEY (`id`),
  ADD KEY `receiver_id` (`receiver_id`),
  ADD KEY `sender_id` (`sender_id`);

ALTER TABLE `user`
  ADD PRIMARY KEY (`id`);


ALTER TABLE `avatar`
  MODIFY `id` int(11) NOT NULL AUTO_INCREMENT;

ALTER TABLE `message`
  MODIFY `id` int(11) NOT NULL AUTO_INCREMENT;

ALTER TABLE `user`
  MODIFY `id` int(11) NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=3;

```

Generates the following schema.prisma:

```prisma
model avatar {
  id         Int    @default(autoincrement()) @id
  avatar_url String
  user_id    Int
  user       user   @relation(fields: [user_id], references: [id])

  @@index([user_id], name: "user_id")
}

model message {
  id                             Int    @default(autoincrement()) @id
  sender_id                      Int
  receiver_id                    Int
  message                        String
  user_message_receiver_idTouser user   @relation("message_receiver_idTouser", fields: [receiver_id], references: [id])
  user_message_sender_idTouser   user   @relation("message_sender_idTouser", fields: [sender_id], references: [id])

  @@index([receiver_id], name: "receiver_id")
  @@index([sender_id], name: "sender_id")
}

model user {
  id                                Int       @default(autoincrement()) @id
  name                              String
  avatar                            avatar[]
  message_message_receiver_idTouser message[] @relation("message_receiver_idTouser")
  message_message_sender_idTouser   message[] @relation("message_sender_idTouser")
}

```

Notice how on model **`avatar`**, the reference to the `user` model is just named `user`. This is what I would expect.

However, the **`message`** model has 2 references to user: **`receiver`** and **`sender`**. Instead of the reference being called **`receiver`** and **`sender`**, they are called `user_message_receiver_idTouser `and  `user_message_sender_idTouser`. 

Is this somehow intentional or a bug?

**Comment by ryands17** [comment, marked as answer]
Jey @ConneXNL 👋 
Yes that is randomly generated by Prisma, and you can rename those to whatever you see fit. Also you can run this command:

```
npx prisma introspect --experimental-reintrospection
```

This will make sure the fields that you have edited remain the same. Let me know if this works.

**Reply by jrmyio** [reply]
Ah, so I just tried to rename **avatar's user_id** to **owner_id**, and it seems that relation actually keeps the name **user**.

I thought prisma did something magic  behind the scenes where it would extract the name '**owner**' from a column called '**owner_id**' but appearently that is not the case.

Is there any easy way to customize how Prisma introspect goes from a column name, to the actual name of the field/relation?
What I would really want is to have the message generate 

senderId, receiverId, sender and receiver (those last 2 being the user relation)

**Reply by janpio** [reply]
Hey, what you are asking for is what we call "Introspection Configuration" - defining how Introspection goes from the database schema to Prisma schema. (Issue related to this are tracked under https://github.com/prisma/prisma/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22topic%3A+introspection-configuration%22) 

Right now that is hardcoded and indeed only uses the target model of a relation to name the relation field itself. Being able to configure that is quite some time out as this is a monumental projects with many different ways to do it.

What you can do right now is to introspect and then to rename them once manually to what you want. With release 2.6.0 future `prisma introspect` invocations should also keep that change, until then you will have to do `prisma introspect --experimental-reintrospection` so the manual changes are not overwritten on the next run of `introspect`.

**Reply by ryands17** [reply]
Unfortunately, there's no way to customize that for now, but you can create a feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) for that and we can look into it :)

**Comment by sir-guevara** [comment]
How can i query the message field and also get send name/receiver name?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
